// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package model_service;
option go_package = "./";

service Model {
  rpc InitModel (InitModelRequest) returns (ModelReply) {}
  rpc ReleaseModel (ReleaseModelRequest) returns (ModelReply) {}
  rpc Inference (InferenceRequest) returns (InferenceReply) {}
}

message InferenceRequest {
  int32 model_enum = 1;

  InferenceInputData data = 2;
  ThresholdParam param = 3;
}

message InferenceInputData{
  int32 width = 1;
  int32 height = 2;
  int32 stride = 3;
  bytes pic_data = 4;
  int64 pic_size = 5;
  bool use_physic_addr = 6;
  int32 sensor_id = 7;
}

message ThresholdParam{
  float threshold = 2;
  float nms_threshold = 3;
}

message InitModelRequest {
  int32 model_enum = 1;
  map<int32, string> model_param = 2; // model type enum: model path
}

message ReleaseModelRequest {
  int32 model_enum = 1;
}

message ModelReply {
  int32 ret_code = 1;
  string err_message = 2;
}

message BBox{
  float x0 = 1;//top left corner x -> model scale
  float y0 = 2;//top left corner y -> model scale
  float x1 = 3;//bottom right corner x -> model scale
  float y1 = 4;//bottom right corner y -> model scale
  float x = 5;//bottom right corner y -> model scale
  float y = 6;//bottom right corner y -> model scale
  float w = 7;//bottom right corner y -> model scale
  float h = 8;//bottom right corner y -> model scale
  float objectness = 9; //物体概率
  float confidence = 10; //tvai 某类别的概率(由模型输出的objectness*confidence得到)或事件概率
  float quality = 11;//图像质量分0-1, 1:高质量图像
  string objname = 12;//objtype的文字描述, 在objtype = OTHERS(_A)的情况下, 可以进行透传. 目的: 支持临时算法改动
  string desc = 13;//json infomation, 目的: 预留, 用于临时情况下将信息以json字段方式输出
}

message InferenceReply {
  repeated BBox bboxes = 1;
  int32 ret_code = 2;
  string err_message = 3;
}

